src/hivpy/sexual_behaviour.py

```diff
@@ -130,3 +144,143 @@ def update_sex_groups(self, population: pd.DataFrame):
+        self.update_rred_age(pop_data)
+        pop_data["rred"] = (self.new_partner_factor *
+                            pop_data["rred_personal"])
+        self.init_rred_adc(pop_data)
+        self.init_rred_diagnosis(pop_data)
+        self.init_rred_population(pop_data)
+        self.init_rred_art_adherence(pop_data)
+        self.init_rred_balance(pop_data)
+
+    def init_rred_art_adherence(self, pop_data):
+        pop_data["rred_art_adherence"] = 1
+
+    def update_rred_art_adherence(self, pop_data):
+        if("art_adherence" in pop_data.columns):
+            indices = selector(pop_data, art_adherence=(operator.lt, self.adherence_threshold))
+            pop_data.loc[indices, "rred_art_adherence"] = self.rred_art_adherence
+
+    def update_rred_age(self, pop_data):
+        over_15s = selector(pop_data, age=(operator.ge, 15))
+        age = pop_data.loc[over_15s, "age"]
+        sex = pop_data.loc[over_15s, "sex"]
+        age_index = self.age_index(age)
+        print(self.age_based_risk)
+        print(np.array(age_index), np.array(sex))
+        pop_data.loc[over_15s, "rred_age"] = self.age_based_risk[age_index, sex]
+        print(pop_data["rred_age"])
+
+    def update_rred_long_term_partnered(self, pop_data):
+        pop_data["rred_long_term_partnered"] = 1  # Unpartnered people
+        if("partnered" in pop_data.columns):
+            partnered_idx = selector(pop_data, partnered=(operator.eq, True))
+            pop_data.loc[partnered_idx, "rred_long_term_partnered"] = self.ltp_risk_factor
+            # This might be more efficient, but is also a bit obscure
+            # = ltp_risk_factor if ltp is true, and 1 if ltp is false.
+            # pop_data["rred_ltp"] = pop_data["ltp"]*self.ltp_risk_factor+(1-pop_data["ltp"])
+
+    def init_rred_personal(self, population, n_pop):
+        p_rred_p = self.sb_data.p_rred_p_dist.sample()
+        population["rred_personal"] = np.ones(n_pop)
+        r = rng.uniform(size=n_pop)
+        mask = r < p_rred_p
+        population.loc[mask, "rred_personal"] = 1e-5
+
+    def init_rred_adc(self, population):
+        population["rred_adc"] = 1.0
+        self.update_rred_adc(population)
+
+    def update_rred_adc(self, population):
+        """Updates risk reduction for AIDS defining condition"""
+        # We don't need the rred_adc==1 condition (since they are all set to rred_adc = 1 to start)
+        # It prevents needless assignments but requires checking more conditions
+        # Not sure which is more efficient or if it matters.
+        indices = selector(population, rred_adc=(operator.eq, 1), HIV_status=(operator.eq, True))
+        population.loc[indices, "rred_adc"] = 0.2
+
+    def init_rred_population(self, pop_data):
+        """Initialise general population risk reduction w.r.t. condomless sex with new partners"""
+        self.rred_population = 1
+
+    def update_rred_population(self, pop_data, date):
```

Comment:
Unless I'm missing something!
```suggestion
    def update_rred_population(self, date):
```

Reply:
Correct!